// vite_react_shadcn_ts/src/pages/Community.tsx
import { useState, useEffect, useRef } from 'react';
import { Header } from '@/components/Header';
import { Button } from '@/components/ui/button';
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from '@/components/ui/card';
import { Tabs, TabsContent, TabsList, TabsTrigger } from '@/components/ui/tabs';
import Feed from '@/components/Feed';
import GroupsSection from '@/components/GroupsSection';
import CreatePostModal from '@/components/CreatePostModal';
import { useTranslation } from 'react-i18next';
import { useAuth } from '@/contexts/AuthContext';
import { useToast } from '@/hooks/use-toast';
import communityService from '@/services/communityService';
import {
  Users,
  Plus,
  TrendingUp,
  ArrowDown
} from 'lucide-react';

const Community = () => {
  const { t } = useTranslation();
  const { isAuthenticated } = useAuth();
  const { toast } = useToast();
  const [posts, setPosts] = useState<any[]>([]);
  const [groups, setGroups] = useState<any[]>([]);
  const [isLoading, setIsLoading] = useState(false);
  const [activeTab, setActiveTab] = useState('feed');
  const groupsRef = useRef<HTMLDivElement>(null);

  // Load posts and groups on component mount
  useEffect(() => {
    loadPosts();
    loadGroups();
  }, []);

  const loadPosts = async () => {
    try {
      setIsLoading(true);
      const response = await communityService.getPosts({ limit: 10 });
      setPosts(response.posts);
    } catch (error) {
      console.error('Failed to load posts:', error);
      // Fall back to mock data
      setPosts(mockPosts);
    } finally {
      setIsLoading(false);
    }
  };

  const loadGroups = async () => {
    try {
      const response = await communityService.getGroups({ limit: 10 });
      setGroups(response.groups);
    } catch (error) {
      console.error('Failed to load groups:', error);
      // Fall back to mock data
      setGroups(mockGroups);
    }
  };

  const handlePostCreated = (newPost: any) => {
    setPosts(prevPosts => [newPost, ...prevPosts]);
    toast({
      title: "Post created!",
      description: "Your travel experience has been shared with the community.",
    });
  };

  const handleJoinGroups = () => {
    setActiveTab('groups');
    // Smooth scroll to groups section
    setTimeout(() => {
      groupsRef.current?.scrollIntoView({ 
        behavior: 'smooth',
        block: 'start'
      });
    }, 100);
  };

  const handleJoinGroup = async (groupId: string | number) => {
    if (!isAuthenticated) {
      toast({
        title: "Authentication Required",
        description: "Please sign in to join groups.",
        variant: "destructive"
      });
      return;
    }

    try {
      await communityService.joinGroup(groupId.toString());
      loadGroups(); // Refresh groups
    } catch (error: any) {
      console.error('Join group error:', error);
    }
  };

  const handleCreateGroup = () => {
    toast({
      title: "Create Group",
      description: "Group creation feature coming soon!",
    });
  };

  // Mock data for fallback
  const mockPosts = [
    {
      id: 1,
      author: {
        name: "Sarah Johnson",
        avatar: "https://images.unsplash.com/photo-1494790108755-2616b612b786",
        username: "@sarahj"
      },
      content: "Just had the most amazing sunset in Santorini! The blue domes and white buildings create such a magical atmosphere. Already planning my next visit! 🌅",
      image: "https://images.unsplash.com/photo-1570077188670-e3a8d69ac5ff",
      location: "Santorini, Greece",
      timestamp: "2 hours ago",
      likes: 127,
      comments: 23,
      shares: 8,
      tags: ["sunset", "santorini", "greece", "travel"]
    },
    {
      id: 2,
      author: {
        name: "Mike Chen",
        avatar: "https://images.unsplash.com/photo-1507003211169-0a1dd7228f2d",
        username: "@mikec"
      },
      content: "Tokyo's street food scene is absolutely incredible! Spent the whole day exploring Shibuya and trying different foods. The ramen here is on another level 🍜",
      image: "https://images.unsplash.com/photo-1613929633558-2448d2d03521",
      location: "Tokyo, Japan",
      timestamp: "5 hours ago",
      likes: 89,
      comments: 15,
      shares: 12,
      tags: ["tokyo", "food", "ramen", "streetfood"]
    },
    {
      id: 3,
      author: {
        name: "Emma Wilson",
        avatar: "https://images.unsplash.com/photo-1438761681033-6461ffad8d80",
        username: "@emmaw"
      },
      content: "Hiking in the Swiss Alps was a dream come true! The views from Matterhorn are absolutely breathtaking. Nature never fails to amaze me 🏔️",
      image: "https://images.unsplash.com/photo-1506905925346-21bda4d32df4",
      location: "Swiss Alps, Switzerland",
      timestamp: "1 day ago",
      likes: 203,
      comments: 34,
      shares: 19,
      tags: ["hiking", "alps", "switzerland", "nature"]
    },
    {
      id: 4,
      author: {
        name: "Alex Rivera",
        avatar: "https://images.unsplash.com/photo-1472099645785-5658abf4ff4e",
        username: "@alexr"
      },
      content: "Exploring the vibrant markets of Marrakech was an incredible sensory experience! The colors, sounds, and aromas are unforgettable. 🇲🇦",
      image: "https://images.unsplash.com/photo-1539650116574-75c0c6d73f6e",
      location: "Marrakech, Morocco",
      timestamp: "3 days ago",
      likes: 156,
      comments: 28,
      shares: 15,
      tags: ["morocco", "markets", "culture", "adventure"]
    }
  ];

  const mockGroups = [
    {
      id: 1,
      name: "Solo Travelers Unite",
      members: 15420,
      description: "Connect with fellow solo travelers and share experiences",
      image: "https://images.unsplash.com/photo-1488646953014-85cb44e25828",
      category: "Solo Travel",
      recentActivity: "New post about safety tips"
    },
    {
      id: 2,
      name: "Budget Backpackers",
      members: 8930,
      description: "Tips and tricks for traveling on a budget",
      image: "https://images.unsplash.com/photo-1553062407-98eeb64c6a62",
      category: "Budget Travel",
      recentActivity: "Discussion about cheap hostels"
    },
    {
      id: 3,
      name: "Photography Enthusiasts",
      members: 12600,
      description: "Share your travel photography and get inspired",
      image: "https://images.unsplash.com/photo-1452421822248-d4c2b47f0c81",
      category: "Photography",
      recentActivity: "Photo contest submissions"
    },
    {
      id: 4,
      name: "Foodie Adventures",
      members: 9850,
      description: "Discover amazing local cuisines around the world",
      image: "https://images.unsplash.com/photo-1555396273-367ea4eb4db5",
      category: "Food & Culture",
      recentActivity: "Recipe sharing from Italy"
    },
    {
      id: 5,
      name: "Adventure Seekers",
      members: 11200,
      description: "For those who love extreme sports and outdoor adventures",
      image: "https://images.unsplash.com/photo-1551524164-6cf2ac2d8c9c",
      category: "Adventure",
      recentActivity: "Planning group hiking trip"
    },
    {
      id: 6,
      name: "Family Travel",
      members: 7650,
      description: "Tips and destinations perfect for family vacations",
      image: "https://images.unsplash.com/photo-1511895426328-dc8714191300",
      category: "Family Travel",
      recentActivity: "Kid-friendly destination ideas"
    }
  ];

  const trendingTopics = [
    { name: "#SustainableTravel", posts: 2340 },
    { name: "#HiddenGems", posts: 1890 },
    { name: "#FoodieTravel", posts: 3210 },
    { name: "#AdventureTravel", posts: 2760 },
    { name: "#TravelTips", posts: 4120 }
  ];

  return (
    <div className="min-h-screen bg-background">
      <Header />
      
      <main className="container mx-auto px-4 py-8">
        {/* Hero Section */}
        <div className="text-center mb-12">
          <h1 className="text-4xl font-bold mb-4 bg-gradient-hero bg-clip-text text-transparent">
            Travel Community
          </h1>
          <p className="text-lg text-muted-foreground mb-8 max-w-2xl mx-auto">
            Connect with fellow travelers, share your experiences, and discover amazing destinations together
          </p>
          
          <div className="flex gap-4 justify-center">
            <CreatePostModal onPostCreated={handlePostCreated}>
              <Button className="bg-gradient-hero hover:opacity-90">
                <Plus className="h-4 w-4 mr-2" />
                Create Post
              </Button>
            </CreatePostModal>
            <Button variant="outline" onClick={handleJoinGroups}>
              <Users className="h-4 w-4 mr-2" />
              Join Groups
              <ArrowDown className="h-4 w-4 ml-2" />
            </Button>
          </div>
        </div>

        <div className="grid lg:grid-cols-4 gap-8">
          {/* Main Content */}
          <div className="lg:col-span-3">
            <Tabs value={activeTab} onValueChange={setActiveTab} className="space-y-6">
              <TabsList className="grid w-full grid-cols-3">
                <TabsTrigger value="feed">Feed</TabsTrigger>
                <TabsTrigger value="groups">Groups</TabsTrigger>
                <TabsTrigger value="create">Create Post</TabsTrigger>
              </TabsList>

              {/* Feed Tab */}
              <TabsContent value="feed">
                <Feed 
                  posts={posts.length > 0 ? posts : mockPosts}
                  onRefresh={loadPosts}
                  isLoading={isLoading}
                />
              </TabsContent>

              {/* Groups Tab */}
              <TabsContent value="groups" ref={groupsRef}>
                <GroupsSection
                  groups={groups.length > 0 ? groups : mockGroups}
                  onJoinGroup={handleJoinGroup}
                  onCreateGroup={handleCreateGroup}
                  isAuthenticated={isAuthenticated}
                />
              </TabsContent>

              {/* Create Post Tab */}
              <TabsContent value="create">
                <Card>
                  <CardHeader>
                    <CardTitle>Share Your Travel Experience</CardTitle>
                    <CardDescription>Tell the community about your latest adventure</CardDescription>
                  </CardHeader>
                  <CardContent className="space-y-6">
                    <div className="text-center py-8">
                      <p className="text-muted-foreground mb-4">
                        Create and share your travel stories with the community
                      </p>
                      <CreatePostModal onPostCreated={handlePostCreated}>
                        <Button className="bg-gradient-hero hover:opacity-90">
                          <Plus className="h-4 w-4 mr-2" />
                          Create New Post
                        </Button>
                      </CreatePostModal>
                    </div>
                  </CardContent>
                </Card>
              </TabsContent>
            </Tabs>
          </div>

          {/* Sidebar */}
          <div className="space-y-6">
            {/* Trending Topics */}
            <Card>
              <CardHeader>
                <CardTitle className="flex items-center gap-2">
                  <TrendingUp className="h-5 w-5" />
                  Trending Topics
                </CardTitle>
              </CardHeader>
              <CardContent>
                <div className="space-y-3">
                  {trendingTopics.map((topic) => (
                    <div key={topic.name} className="flex justify-between items-center">
                      <span className="font-medium text-primary hover:underline cursor-pointer">
                        {topic.name}
                      </span>
                      <span className="text-sm text-muted-foreground">
                        {topic.posts.toLocaleString()}
                      </span>
                    </div>
                  ))}
                </div>
              </CardContent>
            </Card>

            {/* Quick Stats */}
            <Card>
              <CardHeader>
                <CardTitle>Community Stats</CardTitle>
              </CardHeader>
              <CardContent>
                <div className="space-y-4">
                  <div className="text-center">
                    <p className="text-2xl font-bold text-primary">50,000+</p>
                    <p className="text-sm text-muted-foreground">Active Travelers</p>
                  </div>
                  <div className="text-center">
                    <p className="text-2xl font-bold text-secondary">120,000+</p>
                    <p className="text-sm text-muted-foreground">Shared Experiences</p>
                  </div>
                  <div className="text-center">
                    <p className="text-2xl font-bold text-accent">180+</p>
                    <p className="text-sm text-muted-foreground">Countries Visited</p>
                  </div>
                </div>
              </CardContent>
            </Card>

            {/* Quick Actions */}
            <Card>
              <CardHeader>
                <CardTitle>Quick Actions</CardTitle>
              </CardHeader>
              <CardContent className="space-y-3">
                <CreatePostModal onPostCreated={handlePostCreated}>
                  <Button variant="outline" className="w-full justify-start">
                    <Plus className="h-4 w-4 mr-2" />
                    Share Experience
                  </Button>
                </CreatePostModal>
                <Button variant="outline" className="w-full justify-start" onClick={handleJoinGroups}>
                  <Users className="h-4 w-4 mr-2" />
                  Browse Groups
                </Button>
                <Button variant="outline" className="w-full justify-start">
                  <TrendingUp className="h-4 w-4 mr-2" />
                  Trending Posts
                </Button>
              </CardContent>
            </Card>
          </div>
        </div>
      </main>
    </div>
  );
};

export default Community;
